core/src/main/java/com/linecorp/armeria/internal/common/stream/DecodedHttpStreamMessage.java

```diff
@@ -125,6 +129,22 @@ private void initialize() {
         }
 
         initialized = true;
+
+        if (publisher instanceof DecodedHttpResponse) {
+            // Make sure to call HttpDecoder.processTrailers() if a downstream has fully consumed HttpData.
+            ((DecodedHttpResponse) publisher).trailers().handleAsync((trailers, cause) -> {
+                // DecodedHttpResponse.trailers() never completes with a cause
+                assert cause == null;
+
+                try {
+                    subscriber.processTrailers(trailers);
+                } catch (Exception e) {
+                    subscriber.handleException(e);
+                }
+                return null;
+            }, executor);
```

Comment:
I think this will stop working if a user wraps `DecodedHttpResponse` with other wrapper before calling `decode()`.